{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.10.16","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# COMP90016 - Assignment 2\nVersion 1. Last edited 14/04/2025\n\n## Semester 1, 2025","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"2c86880f7dbd4ba64c6de52c2385d902","grade":false,"grade_id":"cell-5cd566ae8d583383","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"NAME = \"Hanning Chen\"\n\nID = \"1434345\"","metadata":{"trusted":true},"outputs":[],"execution_count":1},{"cell_type":"markdown","source":"\nThis assignment should be completed by each student individually. Make sure you read this entire document, and ask for help if anything is not clear. Any changes or clarifications to this document will be announced via the LMS.\n\nPlease make sure you review the University's rules on academic integrity: https://academicintegrity.unimelb.edu.au/\n\nYou submission must be your own work. Do not copy material from other students, from the internet or from AI tools. \n\nYour completed notebook file containing all your answers will be turned in via Canvas. Please also submit an HTML file with the output cleared.\n\nTo complete the assignment, finish the tasks in this notebook.\n\nThe tasks are a combination of writing your own code, interpreting the results and answering related short-answer questions.\n\nIn some cases, we have provided test input and test output that you can use to try out your solutions. These tests are just samples and are **not** exhaustive - they may warn you if you've made a mistake, but they are not guaranteed to. It's up to you to decide whether your code is correct.\n\n**Remember to save your work early and often.**","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"750af8623519a4ae9a7ef2940b667978","grade":false,"grade_id":"cell-ea0e1a6df37e43f5","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"## Marking\n\nCells that must be completed to receive marks are clearly labelled. Some cells are code cells, in which you must complete the code to solve a problem. Others are markdown cells, in which you must write your answers to short-answer questions. \n\nCells that must be completed to receive marks are labelled like this:\n\n`# -- GRADED CELL (1 mark) - complete this cell --`\n\nSome graded cells are code cells, in which you must complete the code to solve a problem. Other graded cells are markdown cells, in which you must write your answers to short-answer questions. \n\nYou will see the following text in graded code cells:\n\n```\n# YOUR CODE HERE\nraise NotImplementedError()\n```\n\n***You must remove the `raise NotImplementedError()` line from the cell, and replace it with your solution.***\n\nOnly add answers to graded cells. If you want to import a library or use a helper function, this must be included in a graded cell.\n\nOnly graded cells will be marked.\n**Don't make changes outside graded cells, and don't add or remove cells from the notebook**.\n\n>Word limits, where stated, will be strictly enforced. Answers exceeding the limit **will not be marked**.\n\n>Run-time limits will be imposed for each coding question. The run-time of a code cell can be calculated by including `%time` at the top of your cell. Cells exceeding the run-time limit **will not be marked**. The run-time limits only apply to test cases that are included in this document.\n\nNo marks are allocated to commenting in your code. We do however, encourage efficient and well-commented code.\n\nThe total marks for the assignment add up to 100, and it will be worth 10% of your overall subject grade.\n\nPart 1: 20 marks\n\nPart 2: 35 marks\n\nPart 3: 30 marks\n\nPart 4: 15 marks","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"f3c99f853e3af8f47e33a9bb725357e8","grade":false,"grade_id":"cell-a654a22c19fdefbe","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"## Submitting\n\nBefore you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\\rightarrow$Run All).\n\nMake sure you fill in any place that says `YOUR CODE HERE` or \"YOUR ANSWER HERE\", as well as your name and student ID at the top of this notebook.\n\n\nYour completed notebook file containing all your answers must be turned in via LMS in `.ipynb` format.\nYou must also submit a copy of this notebook in `html` format with the output cleared.\nYou can do this by using the `clear all output` option in the menu.\n\nYour submission should include **only two** files with names formatted as: **Assignment_2.ipynb** and **Assignment_2.html**","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"a223c9a6a8b876260f675f0b76371345","grade":false,"grade_id":"cell-a8f3f9ddb8997b46","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"## Overview\n\nIn this assignment, you will answer questions about sequence alignment and variant calling.\n\nYou will use the `pysam` and `biopython` libraries in your functions. You may want to refer to sections of the documentation for these tools for additional help. Additional to `pysam`, `biopython` and standard Python 3 functions and methods, you may also use any other library we have used in Computational Genomics including `collections`, `numpy`, `pandas`, `math`, `itertools`, `seaborn` and `matplotlib`.\n\nYou will also use the following command-line tools: `BLAST`, `BWA` and `Samtools`. \n\n\nThe following conda command was used to create the binder :\n\n```conda create -n cg_assignment2 bwa samtools pysam biopython matplotlib seaborn blast numpy pandas```\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"e7f29cc0f4c738b754ebd7004fe9079f","grade":false,"grade_id":"cell-231ce2cdcc452d8d","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"## Setup","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"b1e9aace2837e55c85c9be2b30a0c389","grade":false,"grade_id":"cell-cb61bd3dd20d0831","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"import os\nimport requests\nfrom IPython.core.display import HTML\n\nversion = 2025\nbase_url = f'https://raw.githubusercontent.com/melbournebioinformatics/COMP90016/{version}'\n\n# Function to get data.\ndef fetch_file(url, outpath='.'):\n    response = requests.get(url)\n    if response.status_code == 200:\n        print('File found!')\n        # Get the filename from the URL\n        filename = os.path.basename(url).split('?', 1)[0]\n        # Construct the filepath using the specified directory and filename\n        filepath = os.path.join(outpath, filename)\n        # Create the directory if it doesn't exist\n        if not os.path.exists(outpath):\n            print(f'Creating output dir: {outpath}')\n            os.makedirs(outpath)\n        # Check if the file already exists in the specified directory\n        if os.path.exists(filepath):\n            print(f'{filename} already exists in {outpath}. Skip download.')\n        else:\n            with open(filepath, 'wb') as f:\n                f.write(response.content)\n                f.close()\n            print(f'Saved to: {filepath}')\n    else:\n        print(f'File not found: Code {response.status_code}')","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"5efc00fed8fe4afd120be8114ac04555","grade":false,"grade_id":"cell-eca7f6df2f0e77fc","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":2},{"cell_type":"code","source":"# Make the notebook pretty\n\nHTML(requests.get(f'{base_url}/common/style/custom.css').text)","metadata":{"trusted":true},"outputs":[{"execution_count":3,"output_type":"execute_result","data":{"text/plain":"<IPython.core.display.HTML object>","text/html":"<style>\nh1 {\n    border: 3.0px solid #333;\n    padding: 12px 18px;\n    background-color: #C8D3FD;\n    color: #333\n}\n\nh2 {\n    padding: 12px 18px;\n    background-color: #C8D3FD;\n    color: #333\n}\n\n.question {\n    color: rgb(0, 96, 100);\n    background: rgb(178, 235, 242);\n    border: solid 1px rgb(77, 208, 225);\n    padding: 10px;\n}\n\n.info {\n    color: rgb(27, 94, 32);\n    background: rgb(200, 230, 201);\n    border: solid 1px rgb(129, 199, 132);\n    padding: 10px;\n}\n</style>\n"},"metadata":{}}],"execution_count":3},{"cell_type":"code","source":"# Fetch assignment data\nf_names = [\"comp90016_assignment_2_S_harrisii_y.fasta\",\n\"comp90016_assignment_2_database.fasta\",\n\"comp90016_assignment_2_dnaA.bam\",\n\"comp90016_assignment_2_dnaA.bam.bai\",\n\"comp90016_assignment_2_dnaA_ref.fasta\",\n\"comp90016_assignment_2_dnaA_ref.fasta.fai\",\n\"comp90016_assignment_2_query.fasta\",\n\"comp90016_assignment_2_reads_R1.fastq.gz\",\n\"comp90016_assignment_2_reads_R2.fastq.gz\"]\n\nfor filename in f_names:     \n    url = f'{base_url}/assignment/02/data/{filename}?raw=true'\n    fetch_file(url, './data/')","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"8bca57220c0dfb8d1e0d6826a29e979b","grade":false,"grade_id":"cell-0664141ba25828c7","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"File found!\nCreating output dir: ./data/\nSaved to: ./data/comp90016_assignment_2_S_harrisii_y.fasta\nFile found!\nSaved to: ./data/comp90016_assignment_2_database.fasta\nFile found!\nSaved to: ./data/comp90016_assignment_2_dnaA.bam\nFile found!\nSaved to: ./data/comp90016_assignment_2_dnaA.bam.bai\nFile found!\nSaved to: ./data/comp90016_assignment_2_dnaA_ref.fasta\nFile found!\nSaved to: ./data/comp90016_assignment_2_dnaA_ref.fasta.fai\nFile found!\nSaved to: ./data/comp90016_assignment_2_query.fasta\nFile found!\nSaved to: ./data/comp90016_assignment_2_reads_R1.fastq.gz\nFile found!\nSaved to: ./data/comp90016_assignment_2_reads_R2.fastq.gz\n"}],"execution_count":4},{"cell_type":"markdown","source":"## Part 1: BLAST","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"3290207b18030dc1eaf5132425408f0b","grade":false,"grade_id":"cell-364cd6139bcc3f74","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"### Setup","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"7db14284183089f6564b2e55eb5852a6","grade":false,"grade_id":"cell-7bb9f56fdeefa80e","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"Histamine receptors are a class of evolutionarily-related proteins that bind to histamine and trigger cellular responses. There are four histamine receptor proteins that are found across multiple species. These are referred to as HRH1, HRH2, HRH3 and HRH4.\n\n**comp90016_assignment_2_database.fasta** is a multi-FASTA file containing histamine receptor protein sequences from several species.\n\n**comp90016_assignment_2_query.fasta** is a FASTA file containing a protein sequence of a fragment of an unknown histamine receptor.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"db2ef842da9efa33027f31032c3def86","grade":false,"grade_id":"cell-9ed8653a6f115a66","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"We will be using command-line BLAST to execute a search for matches to the query sequence in the database of histamine receptor sequences.\n\nIf you would like to use your local computer, BLAST can be installed via [bioconda](https://anaconda.org/bioconda/blast) or downloaded [here](https://www.ncbi.nlm.nih.gov/books/NBK279671/). Otherwise, BLAST is pre-installed on the subject Binder. \n\n**Execute the following commands to complete the search.**\n\n```\nmkdir -p blastdb/\ncp data/comp90016_assignment_2_database.fasta blastdb/comp90016_assignment_2_database.fasta\n\n# Build a custom BLAST database using the histamine receptor protein sequences in the blast directory.\nmakeblastdb -dbtype prot -in blastdb/comp90016_assignment_2_database.fasta -out blastdb/histamine_receptor_db\n\n# Execute a BLAST search against the custom database with the unknown protein sequence as the query.\nblastp -query data/comp90016_assignment_2_query.fasta -db blastdb/histamine_receptor_db -outfmt \"7 std ppos\"\n```","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"d64f1f6d26e59fe9b528704fe7b83b17","grade":false,"grade_id":"cell-f53aec2e1fb97108","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"### Questions\nIn the cells below, complete the following tasks:","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"70bd779c44fd9b157be2f3613cbfab5c","grade":false,"grade_id":"cell-45278df4fd8f81a3","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 1.1 </h3>\n\n(5 marks, max 50 words)\n\nBased on the results of the BLAST search, predict which type of histamine receptor the query protein sequence fragment originated from (HRH1, HRH2, HRH3 or HRH4). Justify your choice.\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"8c82ed427b3cf37a1849734c15a252da","grade":false,"grade_id":"Q-1-1","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\nThe query protein-sequence most likely originates from HRH3 because of \n - The top ranked sequences are HRH3 with very small E-values - close to 0 and high sequence identity - above 80%;\n - HRH1 and HRH4 are much weaker with relatively high Evalues and low sequence identity;\n - HRH2 does not show up in the table.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 1.2 </h3>\n\n(5 marks, max 50 words)\n\nInterpret the E-value of the top scoring BLAST hit. How confident are you in your BLAST results?\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"77537257d1dde036a421ae70e6d25cd8","grade":false,"grade_id":"Q-1-2","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\nThe top-hit E-value is 1.72e-48 (HRH3_CAVPO) implying the expected number of matches with a similar score that can occur by random chance in the database is approximately 1.72e-48 $\\approx 0$. Such an extremely small e-value indicates the match is highly significant and is not due to randomness.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 1.3 </h3>\n\n(5 marks, max 50 words)\n\nSuppose you were provided with a nucleotide query sequence as a FASTA file (filename: comp90016_assignment_2_query_nucleotide.fasta). Modify the BLAST command provided above so that it would take the nucleotide query sequence as an input. Use the BLOSUM-80 substitution matrix. Please use the same output format and the same custom database used previously. Note that you are not required to execute this command.\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"e1a2b535d5740b3307a7b877732ed2a2","grade":false,"grade_id":"Q-1-3","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\nWe do not need to change the second line,  makeblastdb $\\cdots$, which builds the protein database. Since we change the query sequence to nucleotide, we need the blastx command as follow\n - blastx -query data/comp90016_assignment_2_query_nucleotide.fasta -db blastdb/histamine_receptor_db -outfmt \"7 std ppos\" -matrix BLOSUM80\n\n, which matrix argument is specified.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 1.4 </h3>\n\n(5 marks, max 50 words)\n\nSuppose you execute a BLAST search and the top-scoring result is a match with 100% identity and 100% query coverage.\nThis search was then repeated using the same parameters, but with additional sequences added to the database so that it was twice the original size. Would the E-value of the top-scoring hit increase, decrease or would it remain the same? Would the run-time of the search increase, decrease or would it remain the same?\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"cf76dc4f58ea3fd9e69d0d343924d00d","grade":false,"grade_id":"Q-1-4","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\n**The E-value of the top-scoring hit will increase**, since E-value is proportional to the database size, given query length and statistical parameters unchanged.\n**The run-tim of search will increase** because BLAST has to search through a larger database. ","metadata":{}},{"cell_type":"markdown","source":"## Part 2: Sequence Alignment","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"af856c7e6630c86c6dbf973def98cc96","grade":false,"grade_id":"cell-460c38f0afe64b68","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"### Setup","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"7436a85fa5f7c4294da8fba6901373fb","grade":false,"grade_id":"cell-fc5904f49fa3959c","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"We begin by importing `Biopython` so that we can take advantage of Bio.Align objects for storing and accessing sequence alignments.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"0329a29b936f5a325c2760382473fc82","grade":false,"grade_id":"cell-fea0b53fc71bc64a","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"from Bio import Align, Seq, SeqRecord","metadata":{"trusted":true},"outputs":[],"execution_count":8},{"cell_type":"markdown","source":"In this section, we will be using amino-acid substitution matrices, encoded as 2D dictionaries. The cell below stores the BLOSUM-62 substitution matrix as a dictionary. The protein sequences used to test your code will only contain the one-letter codes of the 20 standard amino-acids.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"5a88951401a742a1c136dcd362378d54","grade":false,"grade_id":"cell-d73ed6f61136388d","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"BLOSUM_62 = {\n    '*':{'*':1,'A':-4,'C':-4,'B':-4,'E':-4,'D':-4,'G':-4,'F':-4,'I':-4,'H':-4,'K':-4,'M':-4,'L':-4,'N':-4,'Q':-4,'P':-4,'S':-4,'R':-4,'T':-4,'W':-4,'V':-4,'Y':-4,'X':-4,'Z':-4},\n    'A':{'*':-4,'A':4,'C':0,'B':-2,'E':-1,'D':-2,'G':0,'F':-2,'I':-1,'H':-2,'K':-1,'M':-1,'L':-1,'N':-2,'Q':-1,'P':-1,'S':1,'R':-1,'T':0,'W':-3,'V':0,'Y':-2,'X':-1,'Z':-1},\n    'C':{'*':-4,'A':0,'C':9,'B':-3,'E':-4,'D':-3,'G':-3,'F':-2,'I':-1,'H':-3,'K':-3,'M':-1,'L':-1,'N':-3,'Q':-3,'P':-3,'S':-1,'R':-3,'T':-1,'W':-2,'V':-1,'Y':-2,'X':-1,'Z':-3},\n    'B':{'*':-4,'A':-2,'C':-3,'B':4,'E':1,'D':4,'G':-1,'F':-3,'I':-3,'H':0,'K':0,'M':-3,'L':-4,'N':3,'Q':0,'P':-2,'S':0,'R':-1,'T':-1,'W':-4,'V':-3,'Y':-3,'X':-1,'Z':1},\n    'E':{'*':-4,'A':-1,'C':-4,'B':1,'E':5,'D':2,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':-2,'L':-3,'N':0,'Q':2,'P':-1,'S':0,'R':0,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':4},\n    'D':{'*':-4,'A':-2,'C':-3,'B':4,'E':2,'D':6,'G':-1,'F':-3,'I':-3,'H':-1,'K':-1,'M':-3,'L':-4,'N':1,'Q':0,'P':-1,'S':0,'R':-2,'T':-1,'W':-4,'V':-3,'Y':-3,'X':-1,'Z':1},\n    'G':{'*':-4,'A':0,'C':-3,'B':-1,'E':-2,'D':-1,'G':6,'F':-3,'I':-4,'H':-2,'K':-2,'M':-3,'L':-4,'N':0,'Q':-2,'P':-2,'S':0,'R':-2,'T':-2,'W':-2,'V':-3,'Y':-3,'X':-1,'Z':-2},\n    'F':{'*':-4,'A':-2,'C':-2,'B':-3,'E':-3,'D':-3,'G':-3,'F':6,'I':0,'H':-1,'K':-3,'M':0,'L':0,'N':-3,'Q':-3,'P':-4,'S':-2,'R':-3,'T':-2,'W':1,'V':-1,'Y':3,'X':-1,'Z':-3},\n    'I':{'*':-4,'A':-1,'C':-1,'B':-3,'E':-3,'D':-3,'G':-4,'F':0,'I':4,'H':-3,'K':-3,'M':1,'L':2,'N':-3,'Q':-3,'P':-3,'S':-2,'R':-3,'T':-1,'W':-3,'V':3,'Y':-1,'X':-1,'Z':-3},\n    'H':{'*':-4,'A':-2,'C':-3,'B':0,'E':0,'D':-1,'G':-2,'F':-1,'I':-3,'H':8,'K':-1,'M':-2,'L':-3,'N':1,'Q':0,'P':-2,'S':-1,'R':0,'T':-2,'W':-2,'V':-3,'Y':2,'X':-1,'Z':0},\n    'K':{'*':-4,'A':-1,'C':-3,'B':0,'E':1,'D':-1,'G':-2,'F':-3,'I':-3,'H':-1,'K':5,'M':-1,'L':-2,'N':0,'Q':1,'P':-1,'S':0,'R':2,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':1},\n    'M':{'*':-4,'A':-1,'C':-1,'B':-3,'E':-2,'D':-3,'G':-3,'F':0,'I':1,'H':-2,'K':-1,'M':5,'L':2,'N':-2,'Q':0,'P':-2,'S':-1,'R':-1,'T':-1,'W':-1,'V':1,'Y':-1,'X':-1,'Z':-1},\n    'L':{'*':-4,'A':-1,'C':-1,'B':-4,'E':-3,'D':-4,'G':-4,'F':0,'I':2,'H':-3,'K':-2,'M':2,'L':4,'N':-3,'Q':-2,'P':-3,'S':-2,'R':-2,'T':-1,'W':-2,'V':1,'Y':-1,'X':-1,'Z':-3},\n    'N':{'*':-4,'A':-2,'C':-3,'B':3,'E':0,'D':1,'G':0,'F':-3,'I':-3,'H':1,'K':0,'M':-2,'L':-3,'N':6,'Q':0,'P':-2,'S':1,'R':0,'T':0,'W':-4,'V':-3,'Y':-2,'X':-1,'Z':0},\n    'Q':{'*':-4,'A':-1,'C':-3,'B':0,'E':2,'D':0,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':0,'L':-2,'N':0,'Q':5,'P':-1,'S':0,'R':1,'T':-1,'W':-2,'V':-2,'Y':-1,'X':-1,'Z':3},\n    'P':{'*':-4,'A':-1,'C':-3,'B':-2,'E':-1,'D':-1,'G':-2,'F':-4,'I':-3,'H':-2,'K':-1,'M':-2,'L':-3,'N':-2,'Q':-1,'P':7,'S':-1,'R':-2,'T':-1,'W':-4,'V':-2,'Y':-3,'X':-1,'Z':-1},\n    'S':{'*':-4,'A':1,'C':-1,'B':0,'E':0,'D':0,'G':0,'F':-2,'I':-2,'H':-1,'K':0,'M':-1,'L':-2,'N':1,'Q':0,'P':-1,'S':4,'R':-1,'T':1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':0},\n    'R':{'*':-4,'A':-1,'C':-3,'B':-1,'E':0,'D':-2,'G':-2,'F':-3,'I':-3,'H':0,'K':2,'M':-1,'L':-2,'N':0,'Q':1,'P':-2,'S':-1,'R':5,'T':-1,'W':-3,'V':-3,'Y':-2,'X':-1,'Z':0},\n    'T':{'*':-4,'A':0,'C':-1,'B':-1,'E':-1,'D':-1,'G':-2,'F':-2,'I':-1,'H':-2,'K':-1,'M':-1,'L':-1,'N':0,'Q':-1,'P':-1,'S':1,'R':-1,'T':5,'W':-2,'V':0,'Y':-2,'X':-1,'Z':-1},\n    'W':{'*':-4,'A':-3,'C':-2,'B':-4,'E':-3,'D':-4,'G':-2,'F':1,'I':-3,'H':-2,'K':-3,'M':-1,'L':-2,'N':-4,'Q':-2,'P':-4,'S':-3,'R':-3,'T':-2,'W':11,'V':-3,'Y':2,'X':-1,'Z':-3},\n    'V':{'*':-4,'A':0,'C':-1,'B':-3,'E':-2,'D':-3,'G':-3,'F':-1,'I':3,'H':-3,'K':-2,'M':1,'L':1,'N':-3,'Q':-2,'P':-2,'S':-2,'R':-3,'T':0,'W':-3,'V':4,'Y':-1,'X':-1,'Z':-2},\n    'Y':{'*':-4,'A':-2,'C':-2,'B':-3,'E':-2,'D':-3,'G':-3,'F':3,'I':-1,'H':2,'K':-2,'M':-1,'L':-1,'N':-2,'Q':-1,'P':-3,'S':-2,'R':-2,'T':-2,'W':2,'V':-1,'Y':7,'X':-1,'Z':-2},\n    'X':{'*':-4,'A':-1,'C':-1,'B':-1,'E':-1,'D':-1,'G':-1,'F':-1,'I':-1,'H':-1,'K':-1,'M':-1,'L':-1,'N':-1,'Q':-1,'P':-1,'S':-1,'R':-1,'T':-1,'W':-1,'V':-1,'Y':-1,'X':-1,'Z':-1},\n    'Z':{'*':-4,'A':-1,'C':-3,'B':1,'E':4,'D':1,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':-1,'L':-3,'N':0,'Q':3,'P':-1,'S':0,'R':0,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':4}}","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"3b97ce67723874b2eaed1c0e34c6abb5","grade":false,"grade_id":"cell-84eff1f1350a2427","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":9},{"cell_type":"markdown","source":"### Questions\nIn the cells below, complete the following tasks:","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"8b952f7989b853e6e33c7c428783e97c","grade":false,"grade_id":"cell-45278df4fd8f81a8","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"<div class=\"info\">\n<h3> Question 2.1 </h3>\n\n(10 marks)\n\n<b>Challenge:</b> Write a Python function to calculate the percentage sequence similarity between two aligned protein sequences. The percentage similarity is defined as the number of positions with similar amino acid residues as a percentage of the total number of positions in the alignment. \n    \nTwo amino acid residues are considered similar based on the input substitution matrix. If two amino acids have a value greater than 0 in their corresponding cell of the input matrix, they are similar. \n\n- [ ] Input: \n    - Pairwise_aln as a `Bio.Align.MultipleSeqAlignment` object that contains only two aligned sequences.\n    - sub_matrix as a 2D dictionary encoding an amino-acid substitution matrix, such as the BLOSUM-62 matrix above.\n- [ ] Output:\n    - Return a floating-point number between 0 and 100. \n    - Round the output to 4 decimal places.\n    - If pairwise_aln does not contain 2 sequences, return None.\n\n</div>","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"0900397e314707a8d0325bbb9d5b1128","grade":false,"grade_id":"Q-2-1-info","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"# GRADED CELL 2.1 (10 marks, max 1 min run-time)\n\ndef percentage_similarity(pairwise_aln, sub_matrix):\n    \"\"\"\n    Calculate the percentage sequence similarity between two aligned protein sequences. \n    Assume that pairwise_aln is an Bio.Align.MultipleSeqAlignment object. \n    Assume sub_matrix is a 2D dictionary encoding an amino-acid substitution matrix. \n    Two amino acid residues are similar if there is a value greater than 0 in the corresponding cell of sub_matrix. \n    The percentage similarity is defined as the number of positions with similar amino acid residues as a percentage of the total number of positions in the alignment.\n    Return a floating point number between 0 and 100.\n    Round the output to 4 decimal places.\n    If pairwise_aln does not contain 2 sequences, return None.\n    \"\"\"\n   \n    # YOUR CODE HERE\n    # Edge case\n    if len(pairwise_aln) != 2:\n        return None\n\n    seq1, seq2 = pairwise_aln[0], pairwise_aln[1]\n    # check both seq1 and seq2 are SeqRecord objects\n    similar_loc = 0.0 # float\n    seq_length = len(seq1)\n    for aa1, aa2 in zip(seq1, seq2):\n        # Skip gaps\n        if aa1 == '-' or aa2 == '-':\n            continue\n        if sub_matrix[aa1][aa2] > 0:\n            similar_loc += 1 \n        \n    \n    percentage = (similar_loc / seq_length) * 100\n    return round(percentage,4)\n        ","metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"8ec2e41bd2f981e7814258a753fdd7bd","grade":false,"grade_id":"Q-2-1","locked":false,"schema_version":3,"solution":true,"task":false},"trusted":true},"outputs":[],"execution_count":10},{"cell_type":"code","source":"# ~~ Test your function in this cell ~~\ndemo_sequence_a = SeqRecord.SeqRecord(Seq.Seq(\"MYWIW\"))\ndemo_sequence_b = SeqRecord.SeqRecord(Seq.Seq(\"IYW--\"))\ndemo_pairwise_prot_aln = Align.MultipleSeqAlignment([demo_sequence_a, demo_sequence_b])\n\nsequence_a = SeqRecord.SeqRecord(Seq.Seq(\"MYGEGEPGGWQDHVTVLATRRHPKWAQAWVSTMPWGYECGFSRAWVHQTPWINV-----VSLSSHEAYGVVAVRHPWEIFSPYEVYAPYVQDTQHHGNPGQFTTSCYPDE\"))\nsequence_b = SeqRecord.SeqRecord(Seq.Seq(\"MYADGEPGAWQDHMTVLAIYWHHKWAHAWVSTMPWSYECGFSRAWVHQTPWINVIRFTQVSLSSRAWYGILAVRHPWEIFSPYDVYAPYVAATQHHGNPGQFSTSCYP--\"))\npairwise_prot_aln = Align.MultipleSeqAlignment([sequence_a, sequence_b])\n\nprint(percentage_similarity(demo_pairwise_prot_aln, BLOSUM_62)) # should return 60.0\n\nprint(percentage_similarity(pairwise_prot_aln, BLOSUM_62))","metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"60.0\n81.8182\n"}],"execution_count":11},{"cell_type":"code","source":"seq1 = demo_sequence_a\nseq2 = demo_sequence_b\nfor aa1,aa2 in zip(seq1, seq2):\n    print(aa1,aa2)","metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"M I\nY Y\nW W\nI -\nW -\n"}],"execution_count":12},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\nimport math","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"ff41d51443599f1c72f3263b8c25ffa4","grade":true,"grade_id":"T-2-1a","locked":true,"points":10,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":13},{"cell_type":"markdown","source":"<div class=\"info\">\n<h3> Question 2.2 </h3>\n\n(10 marks)\n\n<b>Challenge:</b> A new scoring system for pairwise protein sequence alignments has been proposed. The score for an alignment is the sum of the scores at each position in the alignment. The score from a match or a mismatch is taken from a given substitution matrix. Gaps are penalised with a gap open and a gap extend penalty. Terminal gaps are treated the same as internal gaps.\n \nWrite a Python function to compute this score from a pairwise protein alignment. \n- [ ] Input: \n    - Assume that pairwise_aln is an `Bio.Align.MultipleSeqAlignment` object. \n    - Assume sub_matrix is a 2D dictionary encoding an amino-acid substitution matrix. \n    - Assume gap_open and gap_extend are **negative** integers. \n- [ ] Output:\n    - Return the score as an integer. \n    - If pairwise_aln does not contain 2 sequences, return None.\n\n</div>\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"ef57ba73649d223cc4fa9ae1ee10fef5","grade":false,"grade_id":"Q-2-2-info","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"# GRADED CELL 2.2 (10 marks, max 1 min run-time)\n\ndef alignment_score(pairwise_aln, sub_matrix, gap_open, gap_extend):\n    \"\"\"\n    Compute the alignment score described above from a pairwise protein alignment. \n    Assume that pairwise_aln is an Bio.Align.MultipleSeqAlignment object. \n    Assume sub_matrix is a 2D dictionary encoding an amino-acid substitution matrix. \n    Assume gap_open and gap_extend are negative integers.\n    Terminal gaps are treated the same as internal gaps.\n    Return the score as an integer. \n    If pairwise_aln does not contain 2 sequences, return None.\n    \"\"\"\n   \n    # YOUR CODE HERE\n    # Edge case\n    if len(pairwise_aln) != 2:\n        return None\n\n    seq1, seq2 = pairwise_aln[0], pairwise_aln[1]\n    score = 0 # int\n    # Whether currently in a gap of either sequences\n    seq1_in_gap = False\n    seq2_in_gap = False\n\n    for aa1, aa2 in zip(seq1, seq2):\n        # there are four cases in total\n        # gap in seq1\n        if aa1 == '-' and aa2 != '-':\n            if not seq1_in_gap:\n                score += gap_open \n                seq1_in_gap = True\n            else:\n                score += gap_extend\n\n        # gap in seq2\n        elif aa1 != '-' and aa2 == '-':\n            if not seq2_in_gap:\n                score += gap_open\n                seq2_in_gap =True\n            else:\n                score += gap_extend\n\n        # both gaps\n        elif aa1 == '-' and aa2 == '-':\n            print(f'According to Question 2.3, aligned-gap positions weigh 0. Such positions have been skipped.')\n            continue\n\n        # match/mismatch\n        elif aa1 != '-' and aa2 != '-':\n            score += sub_matrix[aa1][aa2]\n            seq1_in_gap = False\n            seq2_in_gap = False\n\n    return int(score)\n            \n","metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"706ccfde015585608e355866f48245bb","grade":false,"grade_id":"Q-2-2","locked":false,"schema_version":3,"solution":true,"task":false},"trusted":true},"outputs":[],"execution_count":14},{"cell_type":"code","source":"# ~~ Test your function in this cell ~~\nprint(alignment_score(demo_pairwise_prot_aln, BLOSUM_62, -12, -3)) # should return 4\n\nprint(alignment_score(pairwise_prot_aln, BLOSUM_62, -12, -3))","metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"4\n461\n"}],"execution_count":15},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"19214b9717049c0b339d812849d742c9","grade":true,"grade_id":"T-2-2a","locked":true,"points":4,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":16},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"449582f9c97c43e212090a3d31cbf3a0","grade":true,"grade_id":"T-2-2b","locked":true,"points":4,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":17},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"a0a24efcb5755fcdcb8ea2f0b2910c9d","grade":true,"grade_id":"T-2-2c","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":18},{"cell_type":"markdown","source":"<div class=\"info\">\n<h3> Question 2.3 </h3>\n\n(10 marks)\n\nThe sum-of-pairs score considers all pairs of characters at each position in an alignment and sums the scores of every pair. For each pair of characters, please score according to the following scoring system:\n \n- Match (A|A) = 0\n- Mismatch (A|T or A|-) = 1\n- Aligned gaps (-|-) = 0\n \n<b>Challenge:</b> The higher the score, the more variability at that position. The most variable region is defined as the continuous region of length reg_len with the highest sum-of-pairs score across all positions in the region. \n\nWrite a Python function to identify the most variable region in a multiple sequence alignment. \n- [ ] Input:\n    - Assume msa is an `Bio.Align.MultipleSeqAlignment` object. \n    - msa could contain DNA, RNA or protein sequences. \n    - Assume reg_len is a positive integer. \n- [ ] Treat gaps the same as other characters. \n- [ ] Output:  \n    - Return a tuple containing the zero-based indices of the first and last position of the most variable region of length reg_len in the MSA. \n    - If there is a tie, return the indices of the region that occurs first in the MSA. \n    - If msa contains fewer than 3 sequences, return None. \n    - If reg_len is greater than the alignment length, return None.\n\n</div>","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"41bd2558d34498f69029e2201d9cd5ab","grade":false,"grade_id":"Q-2-3-info","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"# GRADED CELL 2.3 (10 marks, max 1 min run-time)\n\ndef variable_region(msa, reg_len):\n    \"\"\"\n    Identify the most variable region in a multiple sequence alignment. \n    The most variable region is defined as the continuous region of length reg_len with the highest sum-of-pairs score across all positions in the region. \n    Treat gaps the same as other characters. \n    Assume msa is an Bio.Align.MultipleSeqAlignment object. \n    Assume reg_len is a positive integer. \n    Return a tuple containing the zero-based indices of the first and last position of the most variable region of length reg_len in the MSA. \n    msa could contain DNA, RNA or protein sequences. \n    If there is a tie, return the indices of the region that occurs first in the MSA. \n    If msa contains fewer than 3 sequences, return None. \n    If reg_len is greater than the alignment length, return None.\n    \"\"\"\n   \n    # YOUR CODE HERE\n    num_seqs = len(demo_msa_rna)\n    alignment_length = msa.get_alignment_length() # or seq_len = len(msa[0])\n\n    # Edge cases\n    if num_seqs < 3 or reg_len > alignment_length:\n        return None\n\n    column_scores = []\n    for pos in range(alignment_length):\n        window = [record[pos] for record in msa] # Column window\n        col_score = 0.0\n        for i in range(num_seqs):\n            for j in range(i + 1, num_seqs):\n                aa1 = window[i]\n                aa2 = window[j]\n                if (aa1 == '-' and aa2 == '-') or aa1 == aa2:\n                    col_score += 0\n                else:\n                    col_score +=1\n\n        column_scores.append(col_score)\n\n    # Find the best windows of reg_len with the highest score\n    max_score = -1\n    max_start = 0\n    for start in range(0, alignment_length - reg_len + 1):\n        window_score = sum(column_scores[start:start+reg_len])\n        # Update if we get a bigger window score\n        if window_score > max_score:\n            max_score = window_score\n            max_start = start\n\n    return (max_start, max_start + reg_len -1)\n        \n    ","metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"bf232bcab47eebd25df52c30a62b5a8f","grade":false,"grade_id":"Q-2-3","locked":false,"schema_version":3,"solution":true,"task":false},"trusted":true},"outputs":[],"execution_count":19},{"cell_type":"code","source":"# ~~ Test your function in this cell ~~\ndemo_sequence_c = SeqRecord.SeqRecord(Seq.Seq(\"CACGAG\"))\ndemo_sequence_d = SeqRecord.SeqRecord(Seq.Seq(\"GCCGAG\"))\ndemo_sequence_e = SeqRecord.SeqRecord(Seq.Seq(\"CGCG-G\"))\ndemo_msa_rna = Align.MultipleSeqAlignment([demo_sequence_c, demo_sequence_d, demo_sequence_e])\n\nsequence_c = SeqRecord.SeqRecord(Seq.Seq(\"AACCCCAUGGUGAAAGCAAUCACAAUGACAUGGUAACCAGGGCUUCUUUUCUUAUUGGACCGCUACGCGUUUUAAGUGAUUAUUUUUAGAGCACUCUGGUAGAACUCUUC\"))\nsequence_d = SeqRecord.SeqRecord(Seq.Seq(\"AACCCC------AAAGCAAUCACAAUGACAUGGUUACCAGUGCUUCUUUUCUUAUUGGAGCGCUACGCGCCUUAAGUGAUUAUUUAAAGAGCACUCUGGUAGAACUCU--\"))\nsequence_e = SeqRecord.SeqRecord(Seq.Seq(\"AAGCCCAUGGUGAAAGCAAUCACAAUGACUAGGUAACCAGUGCUUCUUUUCUUAAUGGAU-GCUACGCGUUUUAAGUGAUUAUUUAGAGAGCACUCUGGUAGAACUCU--\"))\nmsa_rna = Align.MultipleSeqAlignment([sequence_c, sequence_d, sequence_e])\n\nprint(variable_region(demo_msa_rna, 2)) # should return (0,1)\n\nprint(variable_region(msa_rna, 10))","metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"(0, 1)\n(2, 11)\n"}],"execution_count":20},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"26e2907fd6d5cb83539ff591e3de59b8","grade":true,"grade_id":"T-2-3a","locked":true,"points":5,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":21},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"28446cd4c2cc1ecff8f83be916ee50c2","grade":true,"grade_id":"T-2-3b","locked":true,"points":3,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":22},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"8834e358b2d9f6f8d2d6fc8c56e65467","grade":true,"grade_id":"T-2-3c","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":23},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 2.4 </h3>\n\n(5 marks, max 100 words)\n\nUse the [online EMBL-EBI implementation of the MUSCLE MSA algorithm](https://www.ebi.ac.uk/jdispatcher/msa/muscle?stype=protein) to align the sequences found in `blastdb/comp90016_assignment_2_database.fasta`. View the results in the 'Alignments' tab.\n\nDescribe the distribution of the regions of similarity across the length of the alignment. Explain how the distribution of regions of similarity can be biologically relevant.\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"a1b6f7f1b61c28213a0374a4af8b07e7","grade":false,"grade_id":"Q-2-4","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\n","metadata":{}},{"cell_type":"markdown","source":"## Part 3: Read alignment","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"0b3c17eeff44cc0d3fa86604ae1b20d4","grade":false,"grade_id":"cell-3b92b603957fa0ab","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"### Setup","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"f3e82a269b7a39e28541a3c8ab644f47","grade":false,"grade_id":"cell-5b5bfcde7169cdf2","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"We begin by importing `pysam` so that we can use it to explore read alignments.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"82f51ffdce0fc15d76eed20e1471798b","grade":false,"grade_id":"cell-ce97d3c79c7d25e9","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"import pysam","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"41621fc92254c0e49cec16ede49efe7c","grade":false,"grade_id":"cell-d9be56a9b0c81404","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":4},{"cell_type":"markdown","source":"**comp90016_assignment_2_S_harrisii_y.fasta** is a FASTA file containing the reference sequence for the Y chromosome of the Tasmanian Devil (*Sarcophilus harrisii*).\n\n**comp90016_assignment_2_reads_R1.fastq.gz** and **comp90016_assignment_2_reads_R2.fastq.gz** are compressed FASTQ files containing short reads.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"ec9264d37c52c778bf5c555c911b2a86","grade":false,"grade_id":"cell-c771806902cdf0c5","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"We will be using the command-line tools `BWA` and `SAMtools` to align the reads to the reference genome and perform some processing.\n\nIf you would like to use your personal computer, you can download them using the following links: [BWA installation](https://github.com/lh3/bwa), [SAMtools installation](http://www.htslib.org/download/).\n\nOr install via bioconda: [BWA](https://anaconda.org/bioconda/bwa) and [samtools](https://anaconda.org/bioconda/samtools).\n\nThese tools are pre-installed on the subject Binder.\n\nExecute the following code block.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"bf189b49b602ad8fc430b42c47701cb8","grade":false,"grade_id":"cell-99ddabe2d88cd0c7","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"%%bash\n# create a working directory to perform the alignment in\nmkdir -p align\n\n# Index the reference sequence into the align directory.\nbwa index -a is ./data/comp90016_assignment_2_S_harrisii_y.fasta -p align/S_harrissi_y\n\n# change to the align directory\ncd align\n\n# Align the reads to the reference\nbwa mem S_harrissi_y ../data/comp90016_assignment_2_reads_R1.fastq.gz ../data/comp90016_assignment_2_reads_R2.fastq.gz > alignment.sam\n\n# Sort the aligned reads and convert to BAM.\nsamtools sort alignment.sam > alignment.bam\n\n# Index the BAM file.\nsamtools index alignment.bam > alignment.bam.bai","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"a0c1cb182c57eb97ad627bad4b2d10bc","grade":false,"grade_id":"cell-28a4ccd43c35b161","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[{"name":"stderr","output_type":"stream","text":"[bwa_index] Pack FASTA... 0.00 sec\n[bwa_index] Construct BWT for the packed sequence...\n[bwa_index] 0.03 seconds elapse.\n0.00 secex] Update BWT... \n0.00 secex] Pack forward-only FASTA... \n0.01 secex] Construct SA from BWT and Occ... \n[main] Version: 0.7.19-r1273\n0016_assignment_2_S_harrisii_y.fasta/S_harrissi_y ./data/comp9\n[main] Real time: 0.061 sec; CPU: 0.051 sec\n[M::bwa_idx_load_from_disk] read 0 ALT contigs\n[M::process] read 8638 sequences (1295700 bp)...\n[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 4307, 0, 0)\norientation FF as there are not enough pairs\norientation FR...nalyzing insert size distribution for \n[M::mem_pestat] (25, 50, 75) percentile: (492, 499, 506)\n computing mean and std.dev: (464, 534) for\n[M::mem_pestat] mean and std.dev: (498.86, 10.00)\nstat] low and high boundaries for proper pairs: (450, 548)\nere are not enough pairsentation RF as th\n[M::mem_pestat] skip orientation RR as there are not enough pairs\n[M::mem_process_seqs] Processed 8638 reads in 0.229 CPU sec, 0.248 real sec\n[main] Version: 0.7.19-r1273\n1.fastq.gz ../data/comp90016_assignment_2_reads_R2.fastq.gznt_2_reads_R\nsecin] Real time: 0.299 sec; CPU: 0.278 \n"}],"execution_count":5},{"cell_type":"markdown","source":"### Questions\nIn the cells below, complete the following tasks:","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"c35f8800262b90a30ef59069ad361d6e","grade":false,"grade_id":"cell-45278df4fd8f81a7","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 3.1 </h3>\n\n(5 marks, max 50 words)\n\nA colleague suggests that instead, you should use the following line of code for the alignment, sorting and compression.\n    \nNote: There is no need to execute this code.\n\n`bwa mem comp90016_assignment_3_S_harrisii_y.fasta comp90016_assignment_3_reads_R1.fastq.gz comp90016_assignment_3_reads_R2.fastq.gz | samtools sort > alignment.bam`\n\nExplain the difference between these commands and the commands you used previously. Is this a better approach?\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"4e2afa800317ac330359be11277cb236","grade":false,"grade_id":"Q-3-1","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\nColleagues' code\n  - Uses a difference reference file rather than BWA-indexed file, less computationally efficient.\n  - Uses pipe to stream alignment results directly to \"samtool sort\" without .sam-file, saving disk space, and good for large dataset.\n","metadata":{}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 3.2 </h3>\n\n(5 marks, max 50 words)\n\nUse SAMtools to determine how many of the input reads were aligned to the reference genome. Give the total number of aligned reads and the percentage of the total reads that were aligned. Include any commands you used in the graded cell.\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"b3863b3aea9488474a11eb9228519743","grade":false,"grade_id":"Q-3-2","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\n%samtools flagstats ./align/alignment.bam \n\nBy the outputs, we conclude that there is 8638 reads in total and the percentage of alignment is 99.72%, that is, 8614 reads.","metadata":{}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 3.3 </h3>\n\n(10 marks, max 100 words)\n    \nSuggest a reason why there are reads in this readset that did not align to the reference genome. Provide supporting evidence using tools and approaches you have learnt during this subject.\n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"1b1f08a4a7281c4c4057f9ce668887cf","grade":false,"grade_id":"Q-3-3","locked":true,"points":10,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"code","source":"%%bash\nsamtools view ./align/alignment.bam | awk '{print $5 $11} | uniq -c","metadata":{"trusted":true},"outputs":[{"name":"stderr","output_type":"stream","text":"bash: line 1: unexpected EOF while looking for matching `''\nnd of file 2: syntax error: unexpected e\n"},{"traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)","Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msamtools view ./align/alignment.bam | awk \u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;124;43mprint $5 $11} | uniq -c\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n","File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2543\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2541\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2542\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2543\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2545\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2546\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2547\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2548\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n","File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/IPython/core/magics/script.py:159\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    158\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 159\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n","File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/IPython/core/magics/script.py:336\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    331\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    332\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    333\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    334\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    335\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 336\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n","\u001b[0;31mCalledProcessError\u001b[0m: Command 'b\"samtools view ./align/alignment.bam | awk '{print $5 $11} | uniq -c\\n\"' returned non-zero exit status 2."],"ename":"CalledProcessError","evalue":"Command 'b\"samtools view ./align/alignment.bam | awk '{print $5 $11} | uniq -c\\n\"' returned non-zero exit status 2.","output_type":"error"}],"execution_count":1},{"cell_type":"markdown","source":"#### -- GRADED CELL (10 marks) - complete this cell --\n\nYOUR ANSWER HERE","metadata":{}},{"cell_type":"markdown","source":"\n### Setup\n\nThe following code blocks relate to questions 3.4 and 3.5","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"b80395f79c9e1e0b7f0395c9c1d07ebb","grade":false,"grade_id":"cell-569c82a1c890bcb6","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"# Import your bam file into a pysam.AlignmentFile object\naln = pysam.AlignmentFile('./align/alignment.bam','rb')","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"ddc5ce06692aa171c89b281d663f029c","grade":false,"grade_id":"cell-de2dddfa934c7204","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[],"execution_count":15},{"cell_type":"code","source":"# You can use fetch() to access individual aligned reads. For example:\nfor read in aln.fetch('s_harrisii_sample', 64, 65):\n     print(read)","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"870949f0336d5b02360dc1c3dfad5452","grade":false,"grade_id":"cell-eaa2a8fa8de791c8","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"s_harrisii_sample-5050\t163\ts_harrisii_sample\t65\t60\t150M\ts_harrisii_sample\t393\t478\tGGGAGGGCTGAGCTAATTTGCTCCCTTCCTACCTTTTTTTGATGGCAAGCCTGATTCATCAATTTTTTTCTTTCTCTGTTTTAGTCAAGTAGGCCTCTAGAGAGGGGAACATCCCACACCCCTTGGTATGTCGTCTCATGATTGTCATTC\tarray('B', [34, 34, 34, 38, 38, 38, 38, 38, 38, 38, 38, 38, 16, 38, 41, 41, 38, 38, 38, 38, 41, 38, 38, 41, 41, 41, 41, 41, 38, 41, 38, 38, 16, 38, 38, 7, 41, 41, 38, 41, 41, 38, 38, 38, 41, 38, 41, 38, 38, 41, 38, 41, 38, 38, 41, 34, 34, 38, 41, 38, 41, 41, 41, 38, 38, 38, 38, 34, 34, 34, 41, 41, 38, 41, 38, 38, 34, 38, 34, 38, 34, 34, 38, 38, 38, 38, 38, 38, 16, 34, 38, 38, 38, 34, 38, 38, 34, 34, 38, 16, 34, 38, 41, 41, 41, 34, 41, 38, 38, 23, 38, 38, 28, 38, 23, 34, 38, 38, 28, 28, 38, 28, 34, 34, 23, 38, 34, 34, 38, 34, 38, 38, 34, 38, 34, 34, 34, 38, 34, 34, 23, 16, 38, 38, 34, 38, 34, 38, 38, 38])\t[('NM', 0), ('MD', '150'), ('MC', '150M'), ('MQ', 60), ('AS', 150), ('XS', 0)]\n"}],"execution_count":19},{"cell_type":"code","source":"# You can use pileup() to access alignment positions. For example:\nfor pileupcolumn in aln.pileup('s_harrisii_sample', 64, 65):\n    # Position in the alignment.\n    print('Reference: ', pileupcolumn.pos)\n    for pileupread in pileupcolumn.pileups:\n        # Position in the read.\n        print('Read: ', pileupread.query_position)","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"954a351e2f3cef892951acf52d321048","grade":false,"grade_id":"cell-4c64423e628bbf45","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"Reference:  64\nRead:  0\nReference:  65\nRead:  1\nReference:  66\nRead:  2\nReference:  67\nRead:  3\nReference:  68\nRead:  4\nReference:  69\nRead:  5\nReference:  70\nRead:  6\nReference:  71\nRead:  7\nReference:  72\nRead:  8\nReference:  73\nRead:  9\nReference:  74\nRead:  10\nReference:  75\nRead:  11\nReference:  76\nRead:  12\nReference:  77\nRead:  13\nReference:  78\nRead:  14\nReference:  79\nRead:  15\nReference:  80\nRead:  16\nReference:  81\nRead:  17\nReference:  82\nRead:  18\nReference:  83\nRead:  19\nReference:  84\nRead:  20\nReference:  85\nRead:  21\nReference:  86\nRead:  22\nReference:  87\nRead:  23\nReference:  88\nRead:  24\nReference:  89\nRead:  25\nReference:  90\nRead:  26\nReference:  91\nRead:  27\nReference:  92\nRead:  28\nReference:  93\nRead:  29\nReference:  94\nRead:  30\nReference:  95\nRead:  31\nReference:  96\nRead:  32\nReference:  97\nRead:  33\nReference:  98\nRead:  34\nReference:  99\nReference:  100\nRead:  36\nReference:  101\nRead:  37\nReference:  102\nRead:  38\nReference:  103\nRead:  39\nReference:  104\nRead:  40\nReference:  105\nRead:  41\nReference:  106\nRead:  42\nReference:  107\nRead:  43\nReference:  108\nRead:  44\nReference:  109\nRead:  45\nReference:  110\nRead:  46\nReference:  111\nRead:  47\nReference:  112\nRead:  48\nReference:  113\nRead:  49\nReference:  114\nRead:  50\nReference:  115\nRead:  51\nReference:  116\nRead:  52\nReference:  117\nRead:  53\nReference:  118\nRead:  54\nReference:  119\nRead:  55\nReference:  120\nRead:  56\nReference:  121\nRead:  57\nReference:  122\nRead:  58\nReference:  123\nRead:  59\nReference:  124\nRead:  60\nReference:  125\nRead:  61\nReference:  126\nRead:  62\nReference:  127\nRead:  63\nReference:  128\nRead:  64\nReference:  129\nRead:  65\nReference:  130\nRead:  66\nReference:  131\nRead:  67\nReference:  132\nRead:  68\nReference:  133\nRead:  69\nReference:  134\nRead:  70\nReference:  135\nRead:  71\nReference:  136\nRead:  72\nReference:  137\nRead:  73\nReference:  138\nRead:  74\nReference:  139\nRead:  75\nReference:  140\nRead:  76\nReference:  141\nRead:  77\nReference:  142\nRead:  78\nReference:  143\nRead:  79\nReference:  144\nRead:  80\nReference:  145\nRead:  81\nReference:  146\nRead:  82\nReference:  147\nRead:  83\nReference:  148\nRead:  84\nReference:  149\nRead:  85\nReference:  150\nRead:  86\nReference:  151\nRead:  87\nReference:  152\nRead:  88\nReference:  153\nRead:  89\nReference:  154\nRead:  90\nReference:  155\nRead:  91\nReference:  156\nRead:  92\nReference:  157\nRead:  93\nReference:  158\nRead:  94\nReference:  159\nRead:  95\nReference:  160\nRead:  96\nReference:  161\nRead:  97\nReference:  162\nRead:  98\nReference:  163\nRead:  99\nReference:  164\nRead:  100\nReference:  165\nRead:  101\nReference:  166\nRead:  102\nReference:  167\nRead:  103\nReference:  168\nRead:  104\nReference:  169\nRead:  105\nReference:  170\nRead:  106\nReference:  171\nRead:  107\nReference:  172\nRead:  108\nReference:  173\nRead:  109\nReference:  174\nRead:  110\nReference:  175\nRead:  111\nReference:  176\nRead:  112\nReference:  177\nRead:  113\nReference:  178\nRead:  114\nReference:  179\nRead:  115\nReference:  180\nRead:  116\nReference:  181\nRead:  117\nReference:  182\nRead:  118\nReference:  183\nRead:  119\nReference:  184\nRead:  120\nReference:  185\nRead:  121\nReference:  186\nRead:  122\nReference:  187\nRead:  123\nReference:  188\nRead:  124\nReference:  189\nRead:  125\nReference:  190\nRead:  126\nReference:  191\nRead:  127\nReference:  192\nRead:  128\nReference:  193\nRead:  129\nReference:  194\nRead:  130\nReference:  195\nRead:  131\nReference:  196\nRead:  132\nReference:  197\nRead:  133\nReference:  198\nRead:  134\nReference:  199\nRead:  135\nReference:  200\nRead:  136\nReference:  201\nRead:  137\nReference:  202\nRead:  138\nReference:  203\nRead:  139\nReference:  204\nRead:  140\nReference:  205\nRead:  141\nReference:  206\nRead:  142\nReference:  207\nRead:  143\nReference:  208\nRead:  144\nReference:  209\nRead:  145\nReference:  210\nRead:  146\nReference:  211\nRead:  147\nReference:  212\nRead:  148\nReference:  213\nRead:  149\n"}],"execution_count":20},{"cell_type":"markdown","source":"### Questions\nIn the cells below, complete the following tasks:","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"5979c726c9c95b4ff04f824b0b70032d","grade":false,"grade_id":"cell-45278df4fd8f81a9","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"<div class=\"question\">\n\n<h3> Question 3.4 </h3>\n\n(5 marks, max 50 words)\n\nThe *LOC116420406* gene is located at positions 46227 to 67944 of the reference sequence (where position 1 is the first base of the reference sequence).\n\nCalculate the mean read depth in the *LOC116420406* gene. Calculate the mean read depth across the whole alignment. Based on your calculations, are the input reads more likely to be from a WES or a WGS experiment? Explain your answer.    \n    \n</div>\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"1c8f5c3e46962a8c0378e0b78c656fde","grade":false,"grade_id":"Q-3-4","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"markdown","source":"#### -- GRADED CELL (5 marks) - complete this cell --\n\nYOUR ANSWER HERE","metadata":{}},{"cell_type":"markdown","source":"<div class=\"info\">\n<h3> Question 3.5 </h3>\n\n(5 marks)\n\n<b>Challenge:</b> Write Python code to plot the read depths for the first 2000 bases of the *LOC116420406* gene as a line graph. \n- Plot the read depth on the Y axis and plot the genomic position on the X axis, using the numbering of the reference sequence. \n- [ ] Add appropriate labels to both the X and the Y axes. \n- [ ] Add an appropriate title. \n- [ ] The plot should be produced inline, in the Jupyter notebook. \n- [ ] Your code does not have to include a function.\n- [ ] You may use the aln pysam.AlignmentFile object as input.\n\n</div>","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"531a705488e273aa8bf7ff9ab0b8eed7","grade":false,"grade_id":"Q-3-5-info","locked":true,"points":5,"schema_version":3,"solution":false,"task":true}}},{"cell_type":"code","source":"# Here's some Jupyter magic to render plots in the notebook\n%matplotlib inline\n\n# You may want to import some additional packages for building and formatting your graphs (non-essential)\nimport numpy as np\nimport matplotlib.pyplot as plt\nimport seaborn","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"fcc08a803d8974cc4cda0c11dee65d70","grade":false,"grade_id":"cell-422fe967733a3f85","locked":true,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# GRADED CELL 3.5 (5 marks, max 1 min run-time)\n\n# Use this cell to make your plot.\n\n# YOUR CODE HERE\nraise NotImplementedError()","metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"2ea9c48a2875d52abc386ec9c11e7013","grade":false,"grade_id":"Q-3-5","locked":false,"schema_version":3,"solution":true,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"## Part 4: Variant calling","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"e985ffef8d4c17303614aabfc817bb83","grade":false,"grade_id":"cell-34c44a8f6b54aaa9","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"### Setup","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"19b6fc7efcd02f44e0956c8b26232c8f","grade":false,"grade_id":"cell-fe3b1097d2747df1","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"Consider the read alignment stored in **data/comp90016_assignment_2_dnaA.bam** (with index file **data/comp90016_assignment_2_dnaA.bam.bai**). Reads have been aligned to the reference sequence **comp90016_assignment_2_dnaA_ref.fasta** (with index file **comp90016_assignment_2_dnaA_ref.fasta.fai**).\n\nWe are interested in finding single nucleotide variants (SNVs) where the sequence of the reads is different to the reference sequence.","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"2db732451feecf406174c88145dcef2f","grade":false,"grade_id":"cell-8812892040dc39e7","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"### Questions\nIn the cells below, complete the following tasks:","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"709514a77559cd9456ab16ca5179e3c7","grade":false,"grade_id":"cell-45278df4fd8f81b7","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"<div class=\"info\">\n<h3> Question 4.1 </h3>\n\n(15 marks)\n\nImplement an SNV caller for haploid organisms as a Python function.\n\nYour function should only call SNVs where there are at least 5 reads aligned to that position, and at least 80% of the aligned bases are a specific non-reference allele.\n\nYour function should only call (and output) SNVs (not indels, CNVs or larger variants).\n\n- [ ] Assume ref is the filename of a FASTA file containing a single, linear sequence.\n- [ ] Assume ref_idx is the filename of an index file for ref with the extension .fai.\n- [ ] Assume alignment is the filename of a BAM file containing reads aligned to the reference sequence ref.\n- [ ] Assume alignment_idx is the filename of an index file for alignment with the extension .bai.\n\nYou are not required to use all input files in your function.\n\nReturn a Pandas dataframe following this VCF-like specification.\n- [ ] Include one row for every SNV called.\n- [ ] Include the following labelled columns: CHROM POS REF ALT QUAL INFO\n- [ ] Set CHROM to the filename of ref (string).\n- [ ] Set POS to the 1-based position of the variant relative to the reference sequence (integer).\n- [ ] Set REF to the reference allele (string).\n- [ ] Set ALT to the non-reference allele (string).\n- [ ] Set QUAL to the average mapping quality of the non-reference allele (floating-point number).\n- [ ] Set INFO to the allele frequency of the non-reference allele (floating-point number). This is defined as the fraction of the aligned bases that are the ALT allele.\n\nIf no SNVs are called, return an empty Pandas dataframe.\n\n\n</div>","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"9cd7c3b53050f6c54fee914489bce160","grade":false,"grade_id":"Q-4-1-info","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"code","source":"# GRADED CELL 4.1 (15 marks, max 1 min run-time)\n\ndef call_SNVs(ref, ref_idx, alignment, alignment_idx):\n    \"\"\"\n    An SNV caller for haploid organisms.\n    Only call SNVs where there are at least 5 reads aligned to that position, and at least 80% of the aligned bases are a specific non-reference allele.\n    Only call (and output) SNVs (not indels or larger variants).\n    Assume ref is the filename of a FASTA file containing a single, linear sequence.\n    Assume ref_idx is the filename of an index file for ref with the extension .fai.\n    Assume alignment is the filename of a BAM file containing reads aligned to the reference sequence ref.\n    Assume alignment_idx is the filename of an index file for alignment with the extension .bai.\n    Return a Pandas dataframe following this VCF-like specification.\n        - Include one row for every SNV called.\n        - Include the following labelled columns: CHROM POS REF ALT QUAL INFO\n        - Set CHROM to the filename of ref (string).\n        - Set POS to the 1-based position of the variant relative to the reference sequence (integer).\n        - Set REF to the reference allele (string).\n        - Set ALT to the non-reference allele (string).\n        - Set QUAL to the average mapping quality of the non-reference allele (floating-point number).\n        - Set INFO to the allele frequency of the non-reference allele (floating-point number). This is defined as the fraction of the aligned bases that are the ATL allele.\n    If no SNVs are called, return an empty Pandas dataframe.\n    \"\"\"\n   \n    # YOUR CODE HERE\n    raise NotImplementedError()","metadata":{"deletable":false,"nbgrader":{"cell_type":"code","checksum":"d2db1b5a4b7b7984a63f78da3e1cb10f","grade":false,"grade_id":"Q-4-1","locked":false,"schema_version":3,"solution":true,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# ~~ Test your function in this cell ~~\nprint(call_SNVs('./data/comp90016_assignment_2_dnaA_ref.fasta',\n                './data/comp90016_assignment_2_dnaA_ref.fasta.fai',\n                './data/comp90016_assignment_2_dnaA.bam',\n                './data/comp90016_assignment_2_dnaA.bam.bai'))\n\n# Output should be a pandas dataframe that looks like this:\nprint(\n\"\"\"\n                                          CHROM  POS REF ALT  QUAL  INFO\n0  ./data/comp90016_assignment_2_dnaA_ref.fasta  141   G   A  60.0   1.0\n\"\"\"\n)","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"564a22addc4598dec3c5d571c0987a76","grade":true,"grade_id":"T-4-1a","locked":true,"points":5,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"ec149f18592f539b0e76b91ab3656714","grade":true,"grade_id":"T-4-1b","locked":true,"points":4,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"55f3e9b73fc6bb226189b131576933d3","grade":true,"grade_id":"T-4-1c","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"0f381cda396f863610c0871e9d7447fa","grade":true,"grade_id":"T-4-1d","locked":true,"points":2,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"ae8c53288a356b6ecc368dab48720e25","grade":true,"grade_id":"T-4-1e","locked":true,"points":1,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"code","source":"# --- AUTOGRADING CELL DO NOT EDIT ----\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"code","checksum":"8171c39101c409c683d00ef11ee2ff94","grade":true,"grade_id":"T-4-1f","locked":true,"points":1,"schema_version":3,"solution":false,"task":false},"trusted":false},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"# END OF ASSIGNMENT","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"9bf90e68f8257f54d6b0189ff5308e3e","grade":false,"grade_id":"cell-5c804f1187504fb9","locked":true,"schema_version":3,"solution":false,"task":false}}},{"cell_type":"markdown","source":"## Submitting\n\nFollow these steps to submit your assignment\n\n1) Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\\rightarrow$Run All).\n\n2) Make sure you have filled in any place that says `YOUR CODE HERE` or \"YOUR ANSWER HERE\"\n\n3) **Clear all output** (in the menubar, select Kernel$\\rightarrow$Restart & Clear Output)\n\n4) Your completed notebook file containing all your answers must be turned in via LMS in `.ipynb` format.\n\n5) You must also submit a copy of this notebook in `html` format with the output cleared.\n\n\nYour submission should include **only two** files with names formatted as: **Assignment_2.ipynb** and **Assignment_2.html**\n","metadata":{"deletable":false,"editable":false,"nbgrader":{"cell_type":"markdown","checksum":"f1186d05eac33225f8595c7f2a01da07","grade":false,"grade_id":"cell-c82c42eca2b5bee4","locked":true,"schema_version":3,"solution":false,"task":false}}}]}